src/tools/rbd_nbd/rbd-nbd.cc

```diff
@@ -792,55 +803,101 @@ static int parse_imgpath(const std::string &imgpath)
   return 0;
 }
 
-static int do_list_mapped_devices()
+static int get_mapped_info(int pid)
 {
-  char path[64];
-  int m = 0;
-  int fd[2];
+  int r;
+  std::string path = "/proc/" + stringify(pid) + "/cmdline";
+  std::ifstream ifs;
+  std::string cmdline;
+  std::vector<const char*> args;
 
-  if (socketpair(AF_UNIX, SOCK_STREAM, 0, fd) == -1) {
-    int r = -errno;
-    cerr << "rbd-nbd: socketpair failed: " << cpp_strerror(-r) << std::endl;
-    return r;
+  ifs.open(path.c_str(), std::ifstream::in);
+  assert (ifs.is_open());
+  ifs >> cmdline;
+
+  for (unsigned i = 0; i < cmdline.size(); i++) {
+    char *arg = &cmdline[i];
+    if (i == 0) {
+      if (strcmp(basename(arg) , "rbd-nbd") != 0) {
+        return -EINVAL;
+      } 
+    } else { 
+      args.push_back(arg);    
+    }
+    
+    while (cmdline[i] != '\0') {
+      i++;
+    }
+  }
+
+  r = parse_args(args);
+  return r;
+}
+
+static int get_map_pid(const std::string& pid_path)
+{
+  int pid = 0;
+  std::ifstream ifs;
+  ifs.open(pid_path.c_str(), std::ifstream::in);
+  if (!ifs.is_open()) {
+    return 0;
   }
+  ifs >> pid;
+  return pid;
+}
+
+static int do_list_mapped_devices()
+{
+  bool r;
+  bool should_print = false;
+  int index = 0;
+  int pid = 0;
+  TextTable tbl;
+
+  tbl.define_column("pid", TextTable::LEFT, TextTable::LEFT);
+  tbl.define_column("pool", TextTable::LEFT, TextTable::LEFT);
+  tbl.define_column("image", TextTable::LEFT, TextTable::LEFT);
+  tbl.define_column("snap", TextTable::LEFT, TextTable::LEFT);
+  tbl.define_column("device", TextTable::LEFT, TextTable::LEFT);
 
   while (true) {
-    snprintf(path, sizeof(path), "/dev/nbd%d", m);
-    int nbd = open_device(path);
-    if (nbd < 0)
+    std::string nbd_path = "/sys/block/nbd" + stringify(index);
+    if(access(nbd_path.c_str(), F_OK) != 0) {
       break;
-    if (ioctl(nbd, NBD_SET_SOCK, fd[0]) != 0)
-      cout << path << std::endl;
-    else
-      ioctl(nbd, NBD_CLEAR_SOCK);
-    close(nbd);
-    m++;
-  }
+    }
+    std::string pid_path = nbd_path + "/pid";
+    pid = get_map_pid(pid_path);
 
-  close(fd[0]);
-  close(fd[1]);
+    if(pid > 0) {
+      r = get_mapped_info(pid);
+      if (r < 0) {
+        index++;
+        continue;
+      } 
+      should_print = true;
+      if (snapname.empty()) {
+        snapname = "-";
```

Comment:
for the result of "rbd showmapped", if an image was mapped without specified snapshot, it would display "-":
`$rbd showmapped
id pool image      snap device
0  rbd  image_test -    /dev/rbd0`

In addition, "device" is the fifth column in the list-mapped result. But when I modified qa/workunits/rbd-nbd.sh, if have to use $4 for blank name, because the value was empty. 

For the two reasons above, I prefer "-" instead of blank name